classes/announcement/DAO.inc.php

```diff
@@ -0,0 +1,133 @@
+<?php
+/**
+ * @file classes/announcement/DAO.inc.php
+ *
+ * Copyright (c) 2014-2021 Simon Fraser University
+ * Copyright (c) 2000-2021 John Willinsky
+ * Distributed under the GNU GPL v3. For full terms see the file docs/COPYING.
+ *
+ * @class announcement
+ *
+ * @brief Read and write announcements to the database.
+ */
+
+namespace PKP\announcement;
+
+use Illuminate\Support\Collection;
+use Illuminate\Support\Facades\App;
+use Illuminate\Support\LazyCollection;
+use PKP\core\EntityDAO;
+use stdClass;
+
+class DAO extends EntityDAO
+{
+    /** @copydoc EntityDAO::$schema */
+    public string $schema = \PKP\Services\PKPSchemaService::SCHEMA_ANNOUNCEMENT;
+
+    /** @copydoc EntityDAO::$table */
+    public string $table = 'announcements';
+
+    /** @copydoc EntityDAO::$settingsTable */
+    public string $settingsTable = 'announcement_settings';
+
+    /** @copydoc EntityDAO::$primarykeyColumn */
+    public string $primaryKeyColumn = 'announcement_id';
+
+    /** @copydoc EntityDAO::$primaryTableColumns */
+    public array $primaryTableColumns = [
+        'id' => 'announcement_id',
+        'assocId' => 'assoc_id',
+        'assocType' => 'assoc_type',
+        'typeId' => 'type_id',
+        'dateExpire' => 'date_expire',
+        'datePosted' => 'date_posted',
+    ];
+
+    /**
+     * @copydoc EntityDAO::newDataObject()
+     */
+    public function newDataObject(): Announcement
+    {
+        return App::make(Announcement::class);
+    }
+
+    /**
+     * @copydoc EntityDAO::get()
+     */
+    public function get(int $id): Announcement
+    {
+        return parent::get($id);
+    }
+
+    /**
+     * Get the total count of rows matching the configured query
+     */
+    public function getCount(Collector $query): int
+    {
+        return $query
+            ->getQueryBuilder()
+            ->select('a.' . $this->primaryKeyColumn)
+            ->get()
+            ->count();
+    }
+
+    /**
+     * Get a list of ids matching the configured query
+     */
+    public function getIds(Collector $query): Collection
+    {
+        return $query
+            ->getQueryBuilder()
+            ->select('a.' . $this->primaryKeyColumn)
+            ->pluck('a.' . $this->primaryKeyColumn);
+    }
+
+    /**
+     * Get a collection of announcements matching the configured query
+     */
+    public function getMany(Collector $query): LazyCollection
+    {
+        $rows = $query
+            ->getQueryBuilder()
+            ->select(['a.*'])
+            ->get();
+
+        return LazyCollection::make(function () use ($rows) {
```

Comment:
Note that the `get` function materializes the full result set into memory, and the LazyCollection only spares us converting from a `stdObject` to our various classes. I wonder whether `each` rather than `get` would perform better when dealing with large result sets...

Reply:
Ah ok, thanks! I will try out `each`.

Reply:
I only found documentation for each as part of [lazily streaming results](https://laravel.com/docs/8.x/queries#streaming-results-lazily) in Laravel 8.x. In Laravel 7, the `each()` method returns a boolean which doesn't seem to be what we want.

I think we might be able to bump to Laravel 8.x. It doesn't look like the requirements have changed ([7.x](https://laravel.com/docs/7.x/installation#server-requirements), [8.x](https://laravel.com/docs/8.x/deployment#server-requirements)), except for PHP going from 7.23 to 7.3.

This is the approach based on the Laravel 8 docs:

```php
public function getMany(Collector $query): LazyCollection
{
    return $query
        ->getQueryBuilder()
        ->lazy()
        ->each(function ($row) {
            yield $this->fromRow($row);
        });
}
```

Does that look right? I guess I'm not totally sure what is happening here. Are we going back to the database with a new query for each row in the result set? Or does it somehow have all of the rows but not materialize them in memory all at once?

I'm happy with either approach.

Reply:
+1 generally for bumping to Laravel 8. I don't think you can `yield` from an anonymous function like that, but anyway, LazyCollection's `map` function should be a tidy way to accomplish the same thing.

Behind the scenes, I think `lazy` relies on `chunk` to get a "page" worth of results at a time. So if e.g. the default chunk size is 100, then a 400-result set would be fetched using 4 separate queries (using `LIMIT` / `OFFSET`).

If we went with this approach we would need to be careful with using these queries to iterate through and modify results en masse -- an `UPDATE` within the first 100 results might cause the next 100 to be served differently. So the more I think about this, the more I don't think we should solve it until we have a specific problem. For now, either a standard `get` feeding a `yield`, or perhaps a LazyCollection that isn't `lazy()` (thus avoiding the chunking problem) with `map`. I'm leaning towards the latter as a clean solution that'll probably pay dividends in permitting more use of LazyCollections.

Reply:
Ok great.

> I don't think you can yield from an anonymous function like that, but anyway, LazyCollection's map function should be a tidy way to accomplish the same thing.

The Laravel docs show an [example](https://laravel.com/docs/8.x/collections#lazy-collections) using an anonymous function. Have I misunderstood something there?